### ISSUE TITLE
New config: More powerful :config- commands

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#2794]
Split off from https://github.com/qutebrowser/qutebrowser/issues/499#issuecomment-306508021 with some minor changes (adding `:config-unset` and `:config-source`, removing `:conf-replace`, see @lamarpavel's answer to the original comment):

---

What I haven't really talked about above is how the commands will look to edit some more complex setting types (like lists or dictionaries) from qutebrowser.

Things which used to be a section (like `searchengines`) will now be a dictionary, and the bindings will be too. There are also various things like `hints -> prev/next-regexes` which will actually be stored as list.

My current idea how to handle this is to add a few new `:config-*` commands, with `:set` being an alias to `:config-set` because `:set` is just a common thing to have.

## Setting

- [x] `:config-set` (aka `:set`) will take a JSON syntax to set the elements, and override any previous settings. Examples:
  - `:set custom-headers '{"X-Answer": "42"}'`

## Resetting a setting

- [x] `:config-unset` (or `:config-reset`) will be needed for various reasons:

- To reset something which was customized via `:set` and now is in `autoconfig.yml` back to the default (i.e. removing it from the `autoconfig.yml`)
- With per-domain settings: To unset a per-domain setting (and go back to either the global customized value, or the default).
- Maybe as `:config-unset --temp` or so to unset all temporarily set options.

## Adding

- [ ] `:config-add` adds a new element to a dictionary or list. Maybe with a `--replace` switch needed to override existing values.

Examples:
  - `:config-add custom-headers X-Answer 42`
  - `:config-add searchengines wiki https://en.wikipedia.org/wiki/{}`
  - `:config-add next-regexes weiter`

## Removing

- [ ] `:config-remove` removes an element by key (dicts) or value (lists).

Examples:
  - `:config-remove custom-headers X-Answer`
  - `:config-remove searchengines wiki`
  - `:config-remove next-regexes weiter`

## Replacing

- [ ] `:config-replace` searches an element as per above, and replaces the value.

Examples:
  - `:config-replace custom-headers X-Answer 23`
  - `:config-replace searchengines wiki https://de.wikipedia.org/wiki/{}`
  - `:config-replace next-regexes weiter [Ww]eiter`

## Source

- [x] `:config-source` reloads the config (see #54). It can take a filename to load that file instead. It should maybe support both `autoconfig.yml` and `config.py` like files.

Examples:
- `:config-source` (reload `autoconfig.yml` and `config.py`)
- `:config-source ~/.config/qutebrowser/night.py` (could be used to set a user CSS and recolor the UI; like a theme)

## Editing

- [x] `:config-edit` to open the `config.py` file in the configured editor.

--

Thoughts? Any other modifications which should be supported?

**Comment by The-Compiler** [comment]
I started a `:config-add`:

```python
    @cmdutils.register(instance='config-commands')
    @cmdutils.argument('option', completion=configmodel.option)
    def config_add(self, option, val1, val2=None):
        """Add a new element to a dict or list in the config.

        Args:
            option: The option to modify.
            val1: The key for dictionaries, or the value to add for lists.
            val2: The value for dictionaries.
        """
        with self._handle_config_error():
            obj = self._config.get_obj(option, mutable=True)

        if isinstance(obj, list):
            if val2 is not None:
                raise cmdexc.CommandError("Only one argument can be given for "
                                          "a list!")
            obj.append(val1)
        elif isinstance(obj, dict):
            if val2 is None:
                raise cmdexc.CommandError("Two arguments are required for a "
                                          "dict!")
            obj[val1] = val2
        else:
            raise cmdexc.CommandError("Can only add to lists/dicts.")
```

but then decided to drop the idea of `:config-add`, `:config-remove` and `:config-replace` for now. It's not clear to me how they'd work with nested dictionaries (like the binding settings), and I'm not sure there's really demand for this.

**Comment by The-Compiler** [comment]
On a second thought, it'd be useful to add search engines, and also to update https://github.com/qutebrowser/qutebrowser/blob/master/misc/userscripts/ripbang - but it doesn't need to be in v1.0.0.

**Comment by The-Compiler** [comment]
As mentioned [on Reddit](https://www.reddit.com/r/qutebrowser/comments/7otpc1/switching_stylesheets_individually/), this would also be useful to have keybindings to add/remove individual stylesheets from `content.user_stylesheet`.

I think we can simply error out for nested structures such as bindings (or just try to add the string the normal way, which will fail).

**Comment by jgkamat** [comment]
@milogert, nope, just saying you're working on it is good, as assigning people to issues in github is a bit difficult. I would start by only working on one of these check-boxes, if this is your first issue, to keep the scope as small as possible. Thanks! :)

**Comment by milogert** [comment]
Sounds good, thanks for the tip. I will work on `:config-add` then for now then.

**Comment by milogert** [comment]
Alright, next up is `:config-remove`!

**Comment by milogert** [comment]
@The-Compiler for remove and replace, were you thinking that I split them like I did with `:config-add`? After writing the method it feels more complicated than you might be going for but much less complicated than the combined version of `:config-add`.

**Comment by The-Compiler** [comment]
I think it'd be better to separate them too - it's more consistent that way, and the signatures (or at least name of the arguments) are slightly different: `:config-remove-list name value` vs. `:config-remove-dict name key`.

**Comment by cancercufasole** [comment]
Won't config cycle -t content.javascript.enabled just cycle the option. If I somehow forget and run it multiple times I might enable js globally.
I might go for an alias for set -t content.javascript.enabled false.
Does your script enable javascript globally for 15 seconds or am I reading it wrong. ( I have no idea how qutebrowser scripts work )

**Comment by jgkamat** [comment]
cancercufasole writes:

> Won't config cycle -t content.javascript.enabled just cycle the option. If I somehow forget and run it multiple times I might enable js globally.

You can add --print to the command to print the current value

> Does your script enable javascript globally for 15 seconds or am I reading it wrong. ( I have no idea how qutebrowser scripts work )

Yes, 15 or 60 seconds. It also handles the case where you start multiple
scripts (and will extend the time you have).

**Comment by The-Compiler** [comment]
It should also be possible for `:config-cycle` to handle dicts somehow (or perhaps in a separate command) - for example, to cycle the default search engine.
